llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp

```diff
@@ -1672,6 +1672,138 @@ static Value *foldSelectInstWithICmpConst(SelectInst &SI, ICmpInst *ICI,
+      if (matchFalseVal(FalseVal, CmpRHS, X, Y, NoOps, OrOps, None) ||
+          matchFalseVal(FalseVal, CmpRHS, X, Y, XorOps, OrOps, None) ||
+          matchFalseVal(FalseVal, CmpRHS, X, Y, AndOps, OrOps, NotRHS))
+        return IC.replaceInstUsesWith(*ISI, FalseVal);
+      else if (match(CmpRHS, m_AllOnes()))
+        return IC.replaceOperand(SI, 1, Null);
+    }
+  }
+
+  // https://alive2.llvm.org/ce/z/5yaUzR
+  // (X|Y)==C ? X&Y : X^Y -> (X^Y)^C : X^Y  or  ~(X^Y)&C : X^Y
+  // (X|Y)==C ? X^Y : X&Y -> (X&Y)^C : X&Y  or  ~(X&Y)&C : X&Y
+  // if C == 0, (X BitwiseOp Y) : FalseVal -> 0:FalseVal
+  // if C == 0, X&Y:X^Y -> 0^(X^Y):X^Y -> X^Y, X^Y:X&Y -> 0^(X&Y):X&Y -> X&Y
+  // X&Y|(X^Y)^C -> (X^Y)^C, X&Y:~(X^Y)&C -> ~(X^Y)&C
+  // X^Y:(X&Y)^C -> (X&Y)^C, X^Y:~(X&Y)&C -> ~(X&Y)&C
+  if (match(CmpLHS, m_Or(m_Value(X), m_Value(Y)))) {
+    if (match(TrueVal, m_c_And(m_Specific(X), m_Specific(Y)))) {
+      if (matchFalseVal(FalseVal, CmpRHS, X, Y, NoOps, XorOps, None) ||
+          matchFalseVal(FalseVal, CmpRHS, X, Y, XorOps, XorOps, None) ||
+          matchFalseVal(FalseVal, CmpRHS, X, Y, AndOps, XorOps, NotInner))
+        return IC.replaceInstUsesWith(*ISI, FalseVal);
+    } else if (match(TrueVal, m_c_Xor(m_Specific(X), m_Specific(Y)))) {
+      if (matchFalseVal(FalseVal, CmpRHS, X, Y, NoOps, AndOps, None) ||
+          matchFalseVal(FalseVal, CmpRHS, X, Y, XorOps, AndOps, None) ||
+          matchFalseVal(FalseVal, CmpRHS, X, Y, AndOps, AndOps, NotInner))
+        return IC.replaceInstUsesWith(*ISI, FalseVal);
+    } else if (match(TrueVal, m_BitwiseLogic(m_Specific(X), m_Specific(Y))) &&
+               match(CmpRHS, m_Zero()))
+      return IC.replaceOperand(SI, 1, Null);
+  }
+
+  // https://alive2.llvm.org/ce/z/dA9Hy-
+  // (X^Y)==C ? X&Y : X|Y -> (X|Y)^C : X|Y  or (X|Y)&~C : X|Y
+  // (X^Y)==C ? X|Y : X&Y -> (X&Y)|C : X&Y  or (X&Y)^ C : X&Y
+  // X^Y==-1 ? X&Y:~(X|Y) -> (X|Y)^-1:~(X|Y)-> ~(X|Y)
+  // X^Y==-1 ? X|Y:  X^Y  -> (X&Y)|-1:  X^Y -> -1:X^Y
+  // X&Y:(X|Y)^C -> (X|Y)^C, X&Y:(X|Y)&~C -> (X|Y)&~C
+  // X|Y:(X&Y)|C -> (X&Y)|C, X|Y:(X&Y)^C  -> (X&Y)^C
+  if (match(CmpLHS, m_Xor(m_Value(X), m_Value(Y)))) {
+    if ((match(TrueVal, m_c_And(m_Specific(X), m_Specific(Y))))) {
+      if (matchFalseVal(FalseVal, CmpRHS, X, Y, XorOps, OrOps, None) ||
+          matchFalseVal(FalseVal, CmpRHS, X, Y, XorOps, OrOps, None) ||
+          matchFalseVal(FalseVal, CmpRHS, X, Y, AndOps, OrOps, NotRHS))
+        return IC.replaceInstUsesWith(*ISI, FalseVal);
+    } else if (match(TrueVal, m_c_Or(m_Specific(X), m_Specific(Y)))) {
+      if (matchFalseVal(FalseVal, CmpRHS, X, Y, OrOps, AndOps, None) ||
+          matchFalseVal(FalseVal, CmpRHS, X, Y, OrOps, AndOps, None) ||
+          matchFalseVal(FalseVal, CmpRHS, X, Y, XorOps, AndOps, None))
+        return IC.replaceInstUsesWith(*ISI, FalseVal);
+      if (match(CmpRHS, m_AllOnes()))
+        return IC.replaceOperand(SI, 1, AllOnes);
+    }
+  }
```

Comment:
This has become fairly complicate (in tracking all the conditions). You might try commenting the conditions with the case they represent to help keep track i.e for the xor case:

```
f (match(CmpLHS, m_Xor(m_Value(X), m_Value(Y)))) {
    if ((match(TrueVal, m_c_And(m_Specific(X), m_Specific(Y))))) {
         // (X ^ Y) == C ? X & Y : (X | Y) ^ C -> (X | Y) ^ C
      if (matchFalseVal(FalseVal, CmpRHS, X, Y, XorOps, OrOps, None) ||
         // (X ^ Y) == C ? X & Y : (X | Y) ^ C -> (X | Y) ^ C !! -- Same as above
          matchFalseVal(FalseVal, CmpRHS, X, Y, XorOps, OrOps, None) ||
         // (X ^ Y) == C ? X ^ Y : (X | Y) & ~C -> (X | Y) & ~C
          matchFalseVal(FalseVal, CmpRHS, X, Y, AndOps, OrOps, NotRHS))
        return IC.replaceInstUsesWith(*ISI, FalseVal);
    } else if (match(TrueVal, m_c_Or(m_Specific(X), m_Specific(Y)))) {
      // (X ^ Y) == C ? X | Y : (X & Y) | C -> (X & Y) | C
      if (matchFalseVal(FalseVal, CmpRHS, X, Y, OrOps, AndOps, None) ||
      // (X ^ Y) == C ? X | Y : (X & Y) | C -> (X & Y) | C !! -- Same as above
          matchFalseVal(FalseVal, CmpRHS, X, Y, OrOps, AndOps, None) ||
      // (X ^ Y) == C ? X | Y : (X & Y) ^ C -> (X & Y) ^ C
          matchFalseVal(FalseVal, CmpRHS, X, Y, XorOps, AndOps, None))
        return IC.replaceInstUsesWith(*ISI, FalseVal);
      (X ^ Y) == -1 ? (X | Y) : FV -> (X ^ Y) == -1 ? -1 : FV
      if (match(CmpRHS, m_AllOnes()))
        return IC.replaceOperand(SI, 1, AllOnes);
    }
  }
```